### PR TITLE
feat: Show empty table groups #2103

### LINKED ISSUE #2103: Show table groups without table rows
#### Wave SDK Version, OS

(e.g. 4.2.0, Windows)

0.26.1, MacOS 

#### Actual behavior

(A clear and concise description of what happened.)
I want to show table groups that may or may not have table rows, But when I provide `rows=[]` to a particular `table_group` it doesn’t visible in the UI.

```
ui.table_group(
    label="Model Name 3",
    rows=[],
    collapsed=False,
)
```

But when I provide a `table_group` with a `table_row` where it has an empty list for `cells` it shows `Invalid Date` text for corresponding empty values where the `table_column`  has `data_type="time"`

```
ui.table_group(
    label="Model Name 2",
    rows=[
        ui.table_row(
            name="model_2_version_3_table_row",
            cells=[],
        )
    ],
    collapsed=False,
),
```

Screenshot 2023-07-26 at 17 50 35

#### Expected behavior

(A clear and concise description of what you expected to happen.)
Table groups should be able to exist even without table rows

#### Steps To Reproduce
sample code

```
from h2o_wave import main, app, Q, ui


@app('/')
async def serve(q: Q):
    q.page['meta'] = ui.form_card(
        box='1 1 5 5',
        items=[
            ui.table(
                name='models_table',
                height="400px",
                columns=[
                    ui.table_column(
                        name="model_version_name_table_column",
                        label="Model Version Name",
                    ),
                    ui.table_column(
                        name="created_date_table_column",
                        label="Created Date",
                        data_type="time",
                    ),
                    ui.table_column(
                        name="last_updated_table_column",
                        label="Last Updated",
                        data_type="time",
                    ),
                ],
                groups=[
                    ui.table_group(
                        label="Model Name 1",
                        rows=[
                            ui.table_row(
                                name="model_1_version_1_table_row",
                                cells=[
                                    "Model Version 1",
                                    "2023/07/26 5:30 pm",
                                    "2022/06/25 04:29 am"
                                ],
                            ),
                            ui.table_row(
                                name="model_1_version_2_table_row",
                                cells=[
                                    "Model Version 2",
                                    "2023/07/26 5:30 pm",
                                    "2022/06/25 04:29 am"
                                ],
                            )
                        ],
                        collapsed=False,
                    ),
                    ui.table_group(
                        label="Model Name 2",
                        rows=[
                            ui.table_row(
                                name="model_2_version_1_table_row",
                                cells=[
                                    "Model Version 1",
                                    "2023/07/26 5:30 pm",
                                    "2022/06/25 04:29 am"
                                ],
                            ),
                            ui.table_row(
                                name="model_2_version_2_table_row",
                                cells=[
                                    "Model Version 2",
                                    "",
                                    ""
                                ],
                            ),
                            ui.table_row(
                                name="model_2_version_3_table_row",
                                cells=[],
                            )
                        ],
                        collapsed=False,
                    ),
                    ui.table_group(
                        label="Model Name 3",
                        rows=[],
                        collapsed=False,
                    )
                ],
            ),
        ]
    )

    await q.page.save()
```

### PR DESCRIPTION
**Pull request by marek-mihok**
**The PR fulfills these requirements:** (check all the apply)

- [x] It's submitted to the `main` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `feat: Add a button #xxx`, where "xxx" is the issue number).
- [x] When resolving a specific issue, the PR description includes `Closes #xxx`, where "xxx" is the issue number.
- [x] If changes were made to `ui` folder, unit tests (`make test`) still pass.
- [x] New/updated tests are included

___

https://github.com/h2oai/wave/assets/23740173/6715b52e-6ca8-4ac9-a976-47a380f1b89b

Closes #2103

### CODE REVIEW

File: ui/src/table.tsx [hunk 1]
```diff
@@ -733,16 +733,17 @@ export const
...
-            const prevGroup = acc[acc.length - 1]
-            prevGroup?.key === group
-              ? prevGroup.count++
-              : acc.push({ key: group, name: group, startIndex: idx, count: 1, isCollapsed: getIsCollapsed(group, expandedRefs.current) })
+          groups = m.groups.reduce((acc, { label }) => {
+            const
+              prevGroup = acc[acc.length - 1],
+              startIndex = prevGroup ? prevGroup.startIndex + prevGroup.count : 0,
+              count = filteredItems.filter(({ group }) => group === label).length
+            acc.push({ key: label, name: label, startIndex, count, isCollapsed: getIsCollapsed(label, expandedRefs.current) })
```
Comment (mturoci) [inline ui/src/table.tsx:741]: This is slow. I wasn't able to figure out why this change is needed either. Can you explain?

File: ui/src/table.tsx [hunk 2]
```diff
@@ -241,11 +241,10 @@ const
     'on-hover': Fluent.CheckboxVisibility.onHover,
     'hidden': Fluent.CheckboxVisibility.hidden,
   },
-  groupByF = function <T extends Dict<any>>(arr: T[], key: S): Dict<any> {
-    return arr.reduce((rv, x: T) => {
-      (rv[x[key]] = rv[x[key]] || []).push(x)
-      return rv
-    }, {} as Dict<any>)
+  groupByF = function <T extends Dict<any>>(items: T[], filteredItems: T[], key: S): Dict<any> {
+    const groupedItems = [...new Set(items.map(item => item[key]))].reduce((arr, val) => { arr[val] = []; return arr }, [] as Dict<any>)
```
[the remark this answers is not in the document]
Comment (marek-mihok) [reply ui/src/table.tsx:245]: Previously I had it implemented like this:
```tsx
    const groupedItems = items.reduce((arr, item) => {
      if (!arr[item.status]) arr[item.status] = []
      return arr
    }, [] as any)
    filteredItems.forEach(item => groupedItems[item[key]].push(item))
    return groupedItems
```

Also there is another alternative using `forEach` only:
```tsx
    const groupedItems: Dict<any> = []
    items.forEach(item => { if (!groupedItems[item[key]]) groupedItems[item[key]] = [] })
    filteredItems.forEach(item => groupedItems[item[key]].push(item))
    return groupedItems
 ```

File: ui/src/table.tsx (hunk 1 above)
Comment (marek-mihok) [reply ui/src/table.tsx:741]: > I wasn't able to figure out why this change is needed either.

With previous implementation `groups` were created based on `filteredItems` which may not contain all available groups from `m.groups`.

File: ui/src/table.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (mturoci) [reply ui/src/table.tsx:245]: While a little faster, still not as fast as it was before as both options iterate over all items, which seems unnecessary.

File: ui/src/table.tsx (hunk 1 above)
Comment (mturoci) [reply ui/src/table.tsx:741]: I see. However, this approach is not good as it yields O(N^2) compared to the original O(N).

File: ui/src/table.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (marek-mihok) [reply ui/src/table.tsx:245]: Let's say you have following `items` and `filteredItems`:
```tsx
const items=[
  {text: 'Aaaa', status: 'Open'},
  {text: 'Bbbb', status: 'Open'},
  {text: 'Cccc', status: 'Closed'},
  {text: 'Dddd', status: 'Open'},
  {text: 'Eeee', status: 'Closed'},
]

const filteredItems=[
  {text: 'Aaaa', status: 'Open'},
  {text: 'Bbbb', status: 'Open'},
  {text: 'Dddd', status: 'Open'},
]
```

Now what you want is to group items by `text`.
The expected output is to get a dictionary of 5 groups:
```tsx
const groups = [
'Aaaa': [{text: 'Aaaa', status: 'Open'}], 
'Bbbb': [{text: 'Bbbb', status: 'Open'}], 
'Cccc': [], 
'Dddd': [{text: 'Dddd', status: 'Open'}], 
'Eeee': []
]
```

How are you supposed to get this output without iterating through all items?

___ 

I've improved my previous solution by iterating through `items` only (no need to iterate through `filteredItems` separately).
```tsx
  groupByF = function <T extends Dict<any>>(items: T[], filteredItems: T[], key: S): Dict<any> {
    return items.reduce((arr, item, idx) => {
      if (!arr[item[key]]) arr[item[key]] = []
      if (idx < filteredItems.length) {
        const groupName = filteredItems[idx][key]
        if (arr[groupName]) arr[groupName].push(filteredItems[idx])
        else arr[groupName] = [filteredItems[idx]]
      }
      return arr
    }, [] as Dict<any>)
  },
```

[the remark this answers is not in the document]
Comment (mturoci) [reply ui/src/table.tsx:245]: > How are you supposed to get this output without iterating through all items?

What's the point of iterating through all the items every single time we need to recalculate the groups when we know the result will be the same as long as `items` did not change?

[the remark this answers is not in the document]
Comment (marek-mihok) [reply ui/src/table.tsx:245]: > as long as `items` did not change

The result won't be the same also if `key` has changed. 

But I understand your point that we don't need to iterate through all items every time we need to recalculate groups in case `key` or `items` are not changed.

In that case I suggest holding a separate list of all groups (including empty ones) which will be recalculated every time `items`, `groupByKey` or `m.groups` has changed. Wdyt?

[the remark this answers is not in the document]
Comment (mturoci) [reply ui/src/table.tsx:245]: > I suggest holding a separate list of all groups (including empty ones) which will be recalculated every time items, groupByKey or m.groups has changed

Yes if "all groups" means the groups related to the currently selected column, not all possible groups across all columns.

[the remark this answers is not in the document]
Comment (marek-mihok) [reply ui/src/table.tsx:245]: > means the groups related to the currently selected column

Yep, this is what I mean.

File: ui/src/table.tsx [hunk 3]
```diff
@@ -727,35 +728,50 @@ export const
...
+            return acc
+          }, {} as Dict<Fluent.IGroup>)
+          : groupByKey !== '*'
+            ? items.reduce((acc, item) => {
+              const group = (item as Fluent.IObjectWithKey & Dict<any>)[groupByKey]
+              if (!acc[group]) acc[group] = ({ key: group, name: group, startIndex: 0, count: 0 })
+              return acc
+            }, {} as Dict<Fluent.IGroup>)
+            : {}
+      }, [m.groups, groupByKey, items]),
```
Comment (mturoci) [inline ui/src/table.tsx:744]: Wouldn't keeping group names only be enough?

```suggestion
      groupNames = React.useMemo(() => {
        return m.groups
          ? m.groups.reduce((acc, { label }) => acc.add(label), new Set<S>())
          : groupByKey !== '*'
            ? (items as Dict<S>[]).reduce((acc, item) => acc.add(item[groupByKey]), new Set<S>())
            : new Set<S>()
      }, [m.groups, groupByKey, items]),
```

File: ui/src/table.tsx [hunk 4]
```diff
@@ -727,35 +728,50 @@ export const
...
-          groups = filteredItems.reduce((acc, { group }, idx) => {
-            const prevGroup = acc[acc.length - 1]
-            prevGroup?.key === group
-              ? prevGroup.count++
-              : acc.push({ key: group, name: group, startIndex: idx, count: 1, isCollapsed: getIsCollapsed(group, expandedRefs.current) })
-            return acc
-          }, [] as Fluent.IGroup[])
+          filteredItems.forEach(({ group }, idx) => {
+            allGroups[group].count === 0
+              ? allGroups[group] = ({ key: group, name: group, startIndex: idx, count: 1, isCollapsed: getIsCollapsed(group, expandedRefs.current) })
```
Comment (mturoci) [inline ui/src/table.tsx:754]: Nit

```suggestion
              ? allGroups[group] = { ...allGroups[group], startIndex: idx, count: 1 }
```

File: ui/src/table.tsx [hunk 5]
```diff
@@ -727,35 +728,50 @@ export const
...
+          : groupByKey !== '*'
+            ? items.reduce((acc, item) => {
+              const group = (item as Fluent.IObjectWithKey & Dict<any>)[groupByKey]
+              if (!acc[group]) acc[group] = ({ key: group, name: group, startIndex: 0, count: 0 })
+              return acc
+            }, {} as Dict<Fluent.IGroup>)
+            : {}
+      }, [m.groups, groupByKey, items]),
       makeGroups = React.useCallback((groupByKey: S, filteredItems: (Fluent.IObjectWithKey & Dict<any>)[]) => {
+        const allGroups: Dict<Fluent.IGroup> = { ...initialGroups }
```
Comment (mturoci) [inline ui/src/table.tsx:746]: If group names are enough:

```suggestion
        const allGroups = [...groupNames].reduce((acc, groupName) => {
          acc[groupName] = { key: groupName, name: groupName, startIndex: 0, count: 0, isCollapsed: getIsCollapsed(groupName, expandedRefs.current) }
          return acc
        }, {} as Dict<Fluent.IGroup>)
```

File: ui/src/table.tsx [hunk 6]
```diff
@@ -727,35 +728,50 @@ export const
...
+          groups = Object.values(allGroups)
         } else {
           let prevSum = 0
           groupedBy = groupByF(filteredItems, groupByKey)
           const
             groupedByKeys = Object.keys(groupedBy),
             groupByColType = m.columns.find(c => c.name === groupByKey)?.data_type
 
-          groups = groupedByKeys.map((key, i) => {
+          groupedByKeys.forEach((key, i) => {
```
Comment (mturoci) [inline ui/src/table.tsx:765]: Isn't there a bug? If the column is a `time` data type, all groups should have the same format, not just non-empty ones. Assuming by code only though, so you gotta try it out. If so, please fix and add accompanying unit test.

**Review by mturoci — CHANGES_REQUESTED** [review]
Looks better now.

File: ui/src/table.tsx (hunk 3 above)
Comment (marek-mihok) [reply ui/src/table.tsx]: It is also an alternative.

Comment (mturoci) [reply ui/src/table.tsx]: But you understand why is this better, right?

Comment (marek-mihok) [reply ui/src/table.tsx]: It takes advantage of set - no need for checking if item exists before inserting.
It takes less memory to store names only.
It is shorter and easier to read.

Anything else?

Comment (mturoci) [reply ui/src/table.tsx]: All correct. The last thing IIRC is that the previous solution mutated `useMemo` objects later on, which could lead into problems later on.

File: ui/src/table.tsx (hunk 6 above)
Comment (marek-mihok) [reply ui/src/table.tsx:762]: Fixed.

**Review by mturoci — APPROVED** [review]
